### ISSUE TITLE
Can not response to my custom webchat from external webservice

### ISSUE THREAD

**Issue by lopejum** [microsoft/BotBuilder-Samples#2460]
Hello

The scenario (basically, I use an Azure Bot as a Bridge between text apps -Teams, Telegram, Skype, custom webchat... and my webservice that produces info as response according the arriving text from those text apps). So:

1. Client writes a text on my custom web-chat
2. Custom webchat requests a token to my Azure bot (using secret key of Direct line Channel)
3. Custom webchat sends the text with token into Authorization header
4. My bot receives the text and re-sends (forward) that text to an external webservice. (I can not wait the response)
5. My external webservice makes some actions with this text
6. The result of the process of that text will be the response I will try to send to my Azure bot for that conversation.

At this point I have: conversationId, userId, recipient, serviceUrl... All of them I have retrieved in my C# bot and sent as params to my webservice.

For every post request to my bot (trying to response to my writen texts by the client) I use activities end-point (depending of the Channel: directline, Telegram, Teams, Azure WebChat) like these:
http://webchat.botframework.../v3/.../conversations/; http://directline.botframework..../v3/.../conversations/; 
https://smba.traffic.....

I send the body as {type, from, recipient, serviceUrl....} In the post request.

My questions:
1. Is It correct/logic the return way to response to my client?
2. If I use the https://mybot.azurewebsites.net/api/messages and I put Authorization header with directline secret key I always get an 500 Error (jwt token error)... 
3. I can not find the way to retrieve the initial token (created in every Channel app) into my C# bot (trying to send as param with the rest of the others param to compose the request post into my webservice)

Whats the best way to response from an external webservice though my bot to the client text app?

Kind regards
Lope

**Comment by stevkan** [comment]
Hi @lopejum, a few questions for you:

- You mention using the bot as a bridge between app services, but you also mention using an external web service that receives/forwards bot responses, takes action on the received text, and eventually sends the response back to the bot.
    - Are you intending the bot to be just a proxy and the external web service to do the heavy lifting?
    - Or, is the external web service meant to be a kind of middleware where the web chat message is received by the bot, is forwarded to the web service, actions are taken, it is returned, and then forwarded on to app service (and vice versa)?
    - Or, something else???
- For every message that is sent from web chat to your bot, how are you determining which app service (Teams, Telegram, etc.) it is going to be sent to? Or, are you trying to send the message from web chat to all those services, at once?
- Are you only wanting to send messages from web chat to an app service and vice versa? Or, are you also trying to send messages between Teams and Telegram, for example?
- "Is It correct/logic the return way to response to my client?"
    -  Can you rephrase this question? I don't fully understand what you are asking.
- "If I use the https://mybot.azurewebsites.net/api/messages and I put Authorization header with directline secret key I always get an 500 Error (jwt token error)..."
    -  The ".../api/messages" API is used by the bot to communicate with the Bot Framework connector service and isn't meant to be called directly. What are you trying to achieve by calling it and passing the Authorization header to it?
- "I can not find the way to retrieve the initial token (created in every Channel app) into my C# bot"
    - I will come back to this after some of the above questions are answered.

I'm curious...why use an external service? Why not use the bot to do you various processing and communicate with the various app services directly from the bot utilizing the different adapters already available?

**Comment by lopejum** [comment]
Hello @stevkan 

First, thanks a lot for your message.

I'll start with the curiosity (we're building an omnichannel intelligent conversational system framework):
1. For commodity, we prefer, in our team, working with Python rather than Node or C#
2. Initially, our architecture was built to work for WhatsApp and telephone call (real-time conversation) channels, based on Twilio API, betwwen our "Bot" and the client
3. At the moment, when you create a Azure Web App Bot, you can not find the Python language, so, we built a custom bot based on Azure Bot Services Python SDK
4. We need to install/deploy our system, both, on-premise and cloud (so, in this way, we can use Azure Web App for Containers to deploy our modules, on cloud, or using our containers on-premises or private cloud -OpenShift-)
5. Finally, we needed to use our "Bot" with every Azure native channel for "text" messages.

So, our final architecture is: 
[ABB] <--> [G] <--> [S] <--> [OS]

Let me explain every module:
[AAB]
[Module A -Azure Bot "Bridge"] It's a simple Azure EchoBot (C#), since we can activate any of native channels (Azure Web-Chat, custom webchat, Telegram, Teams...). This bot receives our text form any text channel and we "forward" this text to the next module [G]

[G]
[Custom Gateway]: A Tornado Web Api, that, that has methods to receive (via websockets) binary info from Twilio telephone call, translate to text, and send it to module [S]. Another service that receives info from Twilio WhatsApp API info (WhatsApp channel), and send it to module [S]. And our simple service that receives text from our [AAB] and redirect to [S]. 

[S]
[Custom Azure Python Bot SDK]: receives the text from [G], makes the hard work of processing the info (call to Azure Maps, Computer Vision, LUIS, QnA Maker, CosmosDB, Storage, etc.). This module does no have any interaction with Internet but the module [G]. And of course, this module is like a custom Azure Composer (hard logic for the dialog)

[OS]
[Other Services]: As explained before: Azure Maps, Computer Vision, LUIS, QnA Maker, CosmosDB, Storage, etc.

For realtime-audio / whatsapp, no problem: We use Twilio API to response from [G].
For input texts, from azure native channels, we needed to response to the client "from [G] crossing [AAB]"... and I couldn't find the way... (Enter from Azure WebChat, Teams, Telegram,... any native channel; and response from DirectLine channel). 

FINALLY, I found a "workaround" and solution for my architecture (create a new proactive service into my [AAB] module for receiving the messages that need to be sent, from my system, to the final client text app):
The idea started from [this blog](https://www.coderepo.blog/posts/how-to-receive-events-bot-framework-sdk-v4-web-api-bot/), that was based in [this info](https://github.com/microsoft/botbuilder-dotnet/issues/787) and implemented and deployed in this original [azure project](https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/16.proactive-messages)

So, that way, I can have any azure native text channel to send & receive messages (proactive) from my custom bot system.

Thanks a lot for your support!!

Best regards
Lope

**Comment by stevkan** [comment]
@lopejum, I'm glad you were able to find a working solution. Closing this issue as resolved.
